### ISSUE TITLE
Create a plugin-specific "$storage" or "$user" object to hold relevant data

### ISSUE THREAD

**Issue by karger** [mavoweb/mavo#879]
Some storage back-ends provide state that could be valuable to mavo authors if they could get access to it.  For example, the firebase plugin provides a unique user id, a name, and an email address for any logged in user, which the mavo app could use to compose personalize the app and to modify the UI based on user roles.   An easy way to provide this state would be to define a special variable, similar to $now, $today, and $hash, that are kept current by the back-end invoking render on built-in pre-defined mavo property.   I'd suggest having a common standard property name---$user or $storage or $plugin seem natural---but to have it be an *object* that contains, as properties, whatever attributes are relevant to the backend.  So for example when using the firebase plugin I could access $plugin.uid, $plugin.email, $plugin.name.   Although it's not clear how I would handle multiple plugins in this case.  Maybe a special variable name for each plugin?  $firebase.uid, etc?

**Comment by DmitrySharabin** [comment]
I believe,  this issue is related to this one: https://github.com/mavoweb/mavo/issues/351

**Comment by karger** [comment]
Agreed but what I am proposing now is more General I think since some plugins might have no notion of a user but other useful properties. It also adopts Lea's suggestion there to have a single top level object to avoid polluting the global namespace

**Comment by LeaVerou** [comment]
Are there any non-user metadata that backends may want to expose? Otherwise this *is* a dupe of #351, since that's not only about universal user properties, it can accommodate backend-specific ones.

**Comment by joyously** [comment]
When I was trying to write a storage backend for use with WordPress, I wanted to make it generic. That means I didn't have a URL to use to connect, but use the domain supplied by the WordPress plugin that loads the Mavo JS, so that domain would be state for the storage backend.
Could there be timing metadata, to implement throttling?

**Comment by LeaVerou** [comment]
> When I was trying to write a storage backend for use with WordPress, I wanted to make it generic. That means I didn't have a URL to use to connect, but use the domain supplied by the WordPress plugin that loads the Mavo JS, so that domain would be state for the storage backend. 

This is what `mv-storage-id` (IIRC) is for, so that authors can short-circuit the URL matching process and just directly select a backend. 
In [Madata](https://madata.dev/) (which will eventually replace that part of Mavo), developers an also directly use the class itself, e.g. `WordPress.from(...)`. If you want, it may be easier to implement the WP backend there (though it's pre-release and under active development, so docs are often lagging behind the actual API…)

> Could there be timing metadata, to implement throttling?

What kind of throttling?

**Comment by joyously** [comment]
I don't remember anything about `mv-storage-id` but then it's been awhile since I've looked at it. I didn't want the author to have to specify the domain, so it would work more easily (the PHP plugin would provide it...this was for self-hosted WordPress).

One use case for throttling is storing comments. Humans don't go as fast as bots, so you can use timing for spam detection and preventing comment flooding.
Another use case for throttling is load balancing for a busy service (like photo or video sharing).
